docs/environment.rst
.. _environment:

=====================
Environment Variables
=====================

ElectrumX takes no command line arguments, instead its behaviour is
controlled by environment variables.  Only a few are required to be
given, the rest will have sensible defaults if not specified.  Many of
the defaults around resource usage are conservative; I encourage you
to review them.

Note: by default the server will only serve to connections from the
same machine.  To be accessible to other users across the internet you
must set **HOST** appropriately; see below.


Required
========

These environment variables are always required:

.. envvar:: COIN

  Must be a :attr:`NAME` from one of the :class:`Coin` classes in
  `lib/coins.py`_.

.. envvar:: DB_DIRECTORY

  The path to the database directory.  Relative paths should be
  relative to the parent process working directory.  This is the
  directory of the `run` script if you use it.

.. envvar:: DAEMON_URL

  A comma-separated list of daemon URLs.  If more than one is provided
  ElectrumX will initially connect to the first, and failover to
  subsequent ones round-robin style if one stops working.

  The generic form of a daemon URL is::

     http://username:password@hostname:port/

  The leading ``http://`` is optional, as is the trailing slash.  The
  ``:port`` part is also optional and will default to the standard RPC
  port for :envvar:`COIN` and :envvar:`NET` if omitted.


For the ``run`` script
======================

The following are required if you use the ``run`` script:

.. envvar:: ELECTRUMX

  The path to the electrumx_server script.  Relative paths should
  be relative to the directory of the ``run`` script.

.. envvar:: USERNAME

  The username the server will run as.


Miscellaneous
=============

These environment variables are optional:

.. envvar:: LOG_FORMAT

  The Python logging `format string
  <https://docs.python.org/3/library/logging.html#logrecord-attributes>`_
  to use.  Defaults to ``%(levelname)s:%(name)s:%(message)s``.

.. envvar:: ALLOW_ROOT

  Set this environment variable to anything non-empty to allow running
  ElectrumX as root.

.. envvar:: NET

  Must be a :envvar:`NET` from one of the **Coin** classes in
  `lib/coins.py`_.  Defaults to ``mainnet``.

.. envvar:: DB_ENGINE

  Database engine for the UTXO and history database.  The default is
  ``leveldb``.  The other alternative is ``rocksdb``.  You will need
  to install the appropriate python package for your engine.  The
  value is not case sensitive.

.. envvar:: HOST

  The host or IP address that the TCP and SSL servers will use when
  binding listening sockets.  Defaults to ``localhost``.  To listen on
  multiple specific addresses specify a comma-separated list.  Set to
  an empty string to listen on all available interfaces (likely both
  IPv4 and IPv6).

.. envvar:: TCP_PORT

  If set ElectrumX will serve TCP clients on
  :envvar:`HOST`\::envvar:`TCP_PORT`.

  .. note:: ElectrumX will not serve TCP connections until it has
            fully caught up with your daemon.

.. envvar:: SSL_PORT

  If set ElectrumX will serve SSL clients on
  :envvar:`HOST`\::envvar:`SSL_PORT`.  If set then
  :envvar:`SSL_CERTFILE` and :envvar:`SSL_KEYFILE` must be defined
  environment variables with values the filesystem paths to those SSL
  files.

  .. note:: ElectrumX will not serve SSL connections until it has
            fully caught up with your daemon.

.. envvar:: RPC_HOST

  The host or IP address that the RPC server will listen on and
  defaults to ``localhost``.  To listen on multiple specific addresses
  specify a comma-separated list.  Servers with unusual networking
  setups might want to specify e.g. ``::1`` or ``127.0.0.1``
  explicitly rather than defaulting to ``localhost``.

  An empty string (normally indicating all interfaces) is interpreted
  as ``localhost``, because allowing access to the server's RPC
  interface to arbitrary connections across the internet is not a good
  idea.

.. envvar:: RPC_PORT

  ElectrumX will listen on this port for local RPC connections.
  ElectrumX listens for RPC connections unless this is explicitly set
  to blank.  The default depends on :envvar:`COIN` and :envvar:`NET`
  (e.g., 8000 for Bitcoin mainnet) if not set, as indicated in
  `lib/coins.py`_.

.. envvar:: DONATION_ADDRESS

  The server donation address reported to Electrum clients.  Defaults
  to empty, which Electrum interprets as meaning there is none.

.. envvar:: BANNER_FILE

  The path to a banner file to serve to clients in Electrum's
  "Console" tab.  Relative file paths must be relative to
  :envvar:`DB_DIRECTORY`.  The banner file is re-read for each new
  client.

  You can place several meta-variables in your banner file, which will be
  replaced before serving to a client.

  + ``$SERVER_VERSION`` is replaced with the ElectrumX version you are
    running, such as ``1.0.10``.
  + ``$SERVER_SUBVERSION`` is replaced with the ElectrumX user agent
    string.  For example, ``ElectrumX 1.0.10``.
  + ``$DAEMON_VERSION`` is replaced with the daemon's version as a
    dot-separated string. For example ``0.12.1``.
  + ``$DAEMON_SUBVERSION`` is replaced with the daemon's user agent
    string.  For example, ``/BitcoinUnlimited:0.12.1(EB16; AD4)/``.
  + ``$DONATION_ADDRESS`` is replaced with the address from the
    :envvar:`DONATION_ADDRESS` environment variable.

  See `here <https://github.com/shsmith/electrumx-banner-updater>`_
  for a script that updates a banner file periodically with useful
  statistics about fees, last block time and height, etc.

.. envvar:: TOR_BANNER_FILE

  As for :envvar:`BANNER_FILE` (which is also the default) but shown
  to incoming connections believed to be to your Tor hidden service.

.. envvar:: ANON_LOGS

  Set to anything non-empty to replace IP addresses in logs with
  redacted text like ``xx.xx.xx.xx:xxx``.  By default IP addresses
  will be written to logs.

.. envvar:: LOG_SESSIONS

  The number of seconds between printing session statistics to the
  log.  The output is identical to the :ref:`sessions` RPC command
  except that :envvar:`ANON_LOGS` is honoured.  Defaults to 3600.  Set
  to zero to suppress this logging.

.. envvar:: REORG_LIMIT

  The maximum number of blocks to be able to handle in a chain
  reorganisation.  ElectrumX retains some fairly compact undo
  information for this many blocks in levelDB.  The default is a
  function of :envvar:`COIN` and :envvar:`NET`; for Bitcoin mainnet it
  is 200.

.. envvar:: EVENT_LOOP_POLICY

  The name of an event loop policy to replace the default asyncio
  policy, if any.  At present only ``uvloop`` is accepted, in which
  case you must have installed the `uvloop`_ Python package.

  If you are not sure what this means leave it unset.

.. envvar:: DROP_CLIENT

  Set a regular expression to disconnect any client based on their
  version string. For example to drop versions from 1.0 to 1.2 use
  the regex ``1\.[0-2]\.\d+``.

.. envvar:: BLACKLIST_URL

  The URL from which to download the blacklist.json file. The blacklist file
  contains a list of IP addresses (and/or hostname globs) to ban.

  URL types currently supported are `http://`, `https://`, and `file://`.
  (Note that `file://` will poll a local file as if it were a remote server.
  Only absolute paths are supported for `file://` URLs, e.g:
  `file:///path/to/file.json`).

  The default blacklist is maintained by the ElectronX developers and the
  Bitcoin Cash server operator community to provide a convenient list of
  currently-active phisher/sybil nodes.  This list is updated frequently.
  This mechanism has been added in April 2019 as of version 1.10.0 due to
  excessive phishing and sybil attacks on the Bitcoin Cash ElectrumX/ElectronX
  network.

  The file is refreshed from the server (or local disk if using `file://`)
  every 5 minutes (configurable via the :envvar:`BLACKLIST_POLL_INTERVAL`
  variable) and any new entries are added to the ban list and old entries no
  longer in the file are removed from the ban list.  Care has been taken so that
  this file's ban lists do clobber whatever you may specify via the rpc 'banip'
  or 'banhost' commands.

  Set this environment variable to the empty string `""` to disable this feature.
  Defaults to: `https://raw.githubusercontent.com/Electron-Cash/electronx-blacklist/master/blacklist.json`.
  (If you would like to contribute to this blacklist please visit:
  `https://github.com/Electron-Cash/electronx-blacklist/`).

.. envvar:: BLACKLIST_POLL_INTERVAL

  The amount of time in seconds to poll BLACKLIST_URL (if non-empty string)
  for the blacklist.json file.  Defaults to 300 seconds (5 minutes).  Setting
  this value below 30 seconds or to a nonsensical value will result in it being
  set to the default (300 seconds).


Resource Usage Limits
=====================

The following environment variables are all optional and help to limit
server resource consumption and prevent simple DoS.

Address subscriptions in ElectrumX are very cheap - they consume about
160 bytes of memory each and are processed efficiently.  I feel the
two subscription-related defaults below are low and encourage you to
raise them.

.. envvar:: MAX_SESSIONS

  The maximum number of incoming connections.  Once reached, TCP and
  SSL listening sockets are closed until the session count drops
  naturally to 95% of the limit.  Defaults to 1,000.

.. envvar:: MAX_SESSIONS_PER_IP

  The maximum number of simultaneous (non-localhost, non-Tor) client connections
  permitted from a single IP address. If a client has more than this many
  connections, subsequent connections will be disallowed. In addition if
  :envvar:`BAN_EXCESSIVE_CONNECTIONS` is `1` (the default), the offending
  client will be automatically banned (unless it's coming from a Tor proxy or
  from localhost, in which case it will never be banned).
  Defaults to 50.

.. envvar:: MAX_SESSIONS_TOR

  The maximum number of simultaneous client connections permitted from
  Tor clients. If there are more than this many connections from the Tor
  IP address (usually, but not always, localhost), subsequent connections will
  be disallowed.  Banning rules are never applied to Tor clients.  This value
  cannot be set to less than 10 as it would interfere with normal server
  operation to do so (since this limit is also used for localhost connections).
  Defaults to 1000.

.. envvar:: BAN_EXCESSIVE_CONNECTIONS

  If 1 (the default), then clients exceeding :envvar:`MAX_SESSIONS_PER_IP` will
  be automatically banned. This ban is never applied to localhost clients and/or
  to Tor clients. Defaults to 1. (Set to the empty string `""` to disable.)

.. envvar:: MAX_SEND

  The maximum size of a response message to send over the wire, in
  bytes.  Defaults to 4,000,000 (except for AuxPoW coins, which default
  to 10,000,000).  Values smaller than 350,000 are taken as 350,000
  because standard Electrum protocol header "chunk" requests are almost
  that large.

  The Electrum protocol has a flaw in that address histories must be
  served all at once or not at all, an obvious avenue for abuse.
  :envvar:`MAX_SEND` is a stop-gap until the protocol is improved to
  admit incremental history requests.  Each history entry is
  approximately 100 bytes so the default is equivalent to a history
  limit of around 40,000 entries, which should be ample for most
  legitimate users.  If you use a higher default bear in mind one
  client can request history for multiple addresses.  Also note that
  the largest raw transaction you will be able to serve to a client is
  just under half of :envvar:`MAX_SEND`, as each raw byte becomes 2
  hexadecimal ASCII characters on the wire.  Very few transactions on
  Bitcoin mainnet are over 500KB in size.

.. envvar:: MAX_SUBS

  The maximum number of address subscriptions across all sessions. When this
  limit is reached, subsequent subscription requests will be met with an RPC
  error until the subscription number gets below this limit again.
  Defaults to 1,000,000.

.. envvar:: MAX_SESSION_SUBS

  The maximum number of address subscriptions permitted to a single
  session.  When this per-session limit is reached, the client will be
  denied subsequent subscriptions. Defaults to 50,000.

.. envvar:: BANDWIDTH_LIMIT

  Per-session periodic bandwidth usage limit in bytes.  This is a soft,
  not hard, limit.  Currently the period is hard-coded to be one hour.
  The default limit value is 8 million bytes.

  Bandwidth usage over each period is totalled, and when this limit is
  exceeded each subsequent request is stalled by sleeping before
  handling it, effectively giving higher processing priority to other
  sessions.

  The more bandwidth usage exceeds this soft limit the longer the next
  request will sleep.  Sleep times are a round number of seconds with
  a minimum of 1.  Each time the delay changes the event is logged.

  Bandwidth usage is gradually reduced over time by "refunding" a
  proportional part of the limit every now and then.

.. envvar:: SESSION_TIMEOUT

  An integer number of seconds defaulting to 600.  Sessions with no
  activity for longer than this are disconnected.  Properly
  functioning Electrum clients by default will send pings roughly
  every 60 seconds.


Peer Discovery
==============

In response to the :func:`server.peers.subscribe` RPC call, ElectrumX
will only return peer servers that it has recently connected to and
verified basic functionality.

If you are not running a Tor proxy ElectrumX will be unable to connect
to onion server peers, in which case rather than returning no onion
peers it will fall back to a hard-coded list.

To give incoming clients a full range of onion servers you will need
to be running a Tor proxy for ElectrumX to use.

ElectrumX will perform peer-discovery by default and announce itself
to other peers.  If your server is private you may wish to disable
some of this.

.. envvar:: PEER_DISCOVERY

  This environment variable is case-insensitive and defaults to
  ``on``.

  If ``on``, ElectrumX will occasionally connect to and verify its
  network of peer servers.

  If ``off``, peer discovery is disabled and a hard-coded default list
  of servers will be read in and served.  If set to ``self`` then peer
  discovery is disabled and the server will only return itself in the
  peers list.

.. envvar:: PEER_DISCOVERY_TOR

  This environment variable is case-insensitive and defaults to
  ``off`` (or ``0``).  It is identical to PEER_DISCOVERY except it affects
  *.onion peers only.

  The default is ``off`` because of the sybil attack vector this represents.

  If ``on``, then ElectrumX will accept peers with *.onion hostnames and attempt
  to verify them, and forward its discoveries to its peers.

  If ``off`` (the default), then ElectrumX will never discover or announce any
  *.onion peers, and will pretend they simply do not exist.

.. envvar:: PEER_ANNOUNCE

  Set this environment variable to empty to disable announcing itself.
  If not defined, or non-empty, ElectrumX will announce itself to
  peers.

  If peer discovery is disabled this environment variable has no
  effect, because ElectrumX only announces itself to peers when doing
  peer discovery if it notices it is not present in the peer's
  returned list.

.. envvar:: FORCE_PROXY

  By default peer discovery happens over the clear internet.  Set this
  to non-empty to force peer discovery to be done via the proxy.  This
  might be useful if you are running a Tor service exclusively and
  wish to keep your IP address private.

.. envvar:: TOR_PROXY_HOST

  The host where your Tor proxy is running.  Defaults to
  ``localhost``.

  If you are not running a Tor proxy just leave this environment
  variable undefined.

.. envvar:: TOR_PROXY_PORT

  The port on which the Tor proxy is running.  If not set, ElectrumX
  will autodetect any proxy running on the usual ports 9050 (Tor),
  9150 (Tor browser bundle) and 1080 (socks).


Server Advertising
==================

These environment variables affect how your server is advertised
by peer discovery (if enabled).

.. envvar:: REPORT_HOST

  The clearnet host to advertise.  If not set, no clearnet host is
  advertised.

.. envvar:: REPORT_TCP_PORT

  The clearnet TCP port to advertise if :envvar:`REPORT_HOST` is set.
  Defaults to :envvar:`TCP_PORT`.  ``0`` disables publishing a TCP
  port.

.. envvar:: REPORT_SSL_PORT

  The clearnet SSL port to advertise if :envvar:`REPORT_HOST` is set.
  Defaults to :envvar:`SSL_PORT`.  ``0`` disables publishing an SSL
  port.

.. envvar:: REPORT_HOST_TOR

  If you wish run a Tor service, this is the Tor host name to
  advertise and must end with ``.onion``.

.. envvar:: REPORT_TCP_PORT_TOR

  The Tor TCP port to advertise.  The default is the clearnet
  :envvar:`REPORT_TCP_PORT`, unless disabled or it is ``0``, otherwise
  :envvar:`TCP_PORT`.  ``0`` disables publishing a Tor TCP port.

.. envvar:: REPORT_SSL_PORT_TOR

  The Tor SSL port to advertise.  The default is the clearnet
  :envvar:`REPORT_SSL_PORT`, unless disabled or it is ``0``, otherwise
  :envvar:`SSL_PORT`.  ``0`` disables publishing a Tor SSL port.

  .. note:: Certificate-Authority signed certificates don't work over
            Tor, so you should set :envvar:`REPORT_SSL_PORT_TOR` to
            ``0`` if yours is not self-signed.


Cache
=====

If synchronizing from the Genesis block your performance might change
by tweaking the cache size.  Cache size is only checked roughly every
minute, so the cache can grow beyond the specified size.  Moreover,
the Python process is often quite a bit fatter than the cache size,
because of Python overhead and also because leveldb consumes a lot of
memory when flushing.  So I recommend you do not set this over 60% of
your available physical RAM:

.. _CACHE:

.. envvar:: CACHE_MB

  The amount of cache, in MB, to use.  The default is 1,200.

  A portion of the cache is reserved for unflushed history, which is
  written out frequently.  The bulk is used to cache UTXOs.

  Larger caches probably increase performance a little as there is
  significant searching of the UTXO cache during indexing.  However, I
  don't see much benefit in my tests pushing this too high, and in
  fact performance begins to fall, probably because LevelDB already
  caches, and also because of Python GC.

  I do not recommend raising this above 2000.

.. _lib/coins.py: https://github.com/kyuupichan/electrumx/blob/master/electrumx/lib/coins.py
.. _uvloop: https://pypi.python.org/pypi/uvloop


docs/changelog.rst
===========
 ChangeLog
===========

.. note:: It is strongly recommended you upgrade to Python 3.7, which
   fixes bugs in asyncio that caused an ever-growing open file count
   and memory consumption whilst serving clients.  Those problems
   should not occur with Python 3.7.

.. note:: The primary difference between this and the official ElectrumX
   is that this version supports any block ordering (including the lexically-
   ordered blocks that can be created/followed by several Bitcoin Cash node
   implementations: ABC, BU, XT, bchd, bcash, Bitprim, and others.). We also
   employ a different anti-phishing strategy than they do, based on a
   community-maintained blacklist.json file as well as manual banning
   to supplement that via the banip command. We also have slightly more
   generous limits to things like MAX_SEND and MAX_SUBS than the defaults
   of ElectrumX.  We also don't implement their newest cost control logic which
   we feel is too restrictive for normal BCH usage of the server.


Version 1.10.1 (21 April 2019)
==============================

* Added PEER_DISCOVERY_TOR environment variable (default off). If on, then
  ElectronX will behave as before this version and always forward/discover
  *.onion peers.  If off, then it will completely ignore *.onion peers and
  never forward them.  This policy change has been implemented because of the
  ease with which sybil attacks are possible if PEER_DISCOVER_TOR is enabled.
  (cculianu)
* Added more stuff to `getenv` rpc command display (cculinau)
* Optimized the ban handling a little bit to refuse peers earlier in the
  lifecycle of peer_discovery if a host is banned. (cculianu)

Version 1.10.0 (14 April 2019)
==============================

* Added banip, unbanip, banhost, unbanhost, listbanned, getenv, and
  session_ip_counts rpc commands (cculianu)
* Disallow multiple servers from same IP to appear in peers (doesn't apply
  to Tor, however) (cculianu)
* Limit client connections per IP to 50 by default (MAX_SESSIONS_PER_IP)
  to prevent the "clients exhausting FDs on server" attack vector (doesn't
  apply to Tor or localhost clients).
  (cculianu)
* Limit client connections per IP for Tor to 1000 by default (MAX_SESSIONS_TOR)
  Note that this limit also applies to clients coming from localhost even if
  not using Tor. (cculianu)
* Added blacklist.json mechanism for downloading a community maintained
  list of bad peers as an anti-phishing, anti-sybil attack measure.
  Use BLACKLIST_URL="" in env to disable, but it is recommended you leave
  it enabled.  We use our own file format for this but the code for this
  subsystem also understand the ElectrumX 1.10.0 format as well. (cculianu)
* Set MAX_SEND to 4000000 (4MB) by default (was originally 1MB which fails
  on some BCH wallets). (cculianu)
* Raised server limit from 250,000 subs to 1,000,000 subs by default
  (corresponds to env var MAX_SUBS) (cculianu)
* Raised default BANDWIDTH_LIMIT from 2MB to 8MB (8000000 bytes).  This is
  because very busy BCH wallets with huge histories would need a higher limit.
  Note the limit is a soft limit in bytes per hour, and it is not catastrophic
  to send a client 8MB in 1 hour. (cculianu)


Version 1.9.4 (7 Feb 2019)
============================
* Synch up to ElectrumX codebase (pulling in relevant & useful commits)
  (cculianu)
* coin additions / updates: SmartCash (rc125), NIX (phamels), Minexcoin
  (joesixpack), BitcoinCash (mblunderburg), Dash (zebra-lucky),
  BitcoinCashRegtest (ezegom), AXE (slowdive), NOR (flo071),
  BitcoinPlus (bushsolo), Myriadcoin (cryptapus), Trezarcoin (ChekaZ), Bitcoin
  Diamond (John Shine), NMC (jeremyrand), Dash (zebra-lucky), PeerCoin
  (peerchemist), BCH testnet (Mark Lundeberg, cculianu), Unitus (ChekaZ)
* close ElectrumX issues: `#554`_, `#653`_, `#655`_, `#684`_, `#713`_
* other minor tweaks (Michael Schmoock, Michael Taborsky)
* increase default MAX_SEND for AuxPow Chains.  Truncate AuxPow for block
  heights covered by a checkpoint.  (jeremyrand)
* tighter RPC param checking (ghost43)
* require aiorpcX 0.10.3

Version 1.8.12 (10 Nov 2018)
============================

* bug fix

Version 1.8.11 (07 Nov 2018)
============================

* require aiorpcX 0.10.1

Version 1.8.10 (05 Nov 2018)
============================

* require aiorpcX 0.10.0
* fix `#632`_
* coin additions / updates: ZelCash (TheTrunk)

Version 1.8.9 (02 Nov 2018)
===========================

* fix `#630`_

Version 1.8.8 (01 Nov 2018)
===========================

* require aiorpcX 0.9.0
* coin additions / updates: decred (dajohi, bolapara), zcash (erasmospunk),
  namecoin (JeremyRand),CivX (turcol), NewYorkCoin (erasmospunk)
* fix `#603`_, `#608`_
* other minor fixes and changes: FMCorz

Version 1.8.7 (13 Sep 2018)
===========================

* require aiorpcX 0.8.1
* fix reorg bug loading blocks from disk (erasmospunk)

Version 1.8.6 (12 Sep 2018)
===========================

* require aiorpcX 0.8.0
* suppress socket.send() errors
* new coin TokenPay (samfiragabriel)
* minor fix: wakiyamap

Version 1.8.5 (18 Aug 2018)
===========================

* require aiorpcX 0.7.3 which contains a couple of bugfixes
* fix `#552`_, `#577`_
* fixed a session limiting bug reported by ghost43
* coin additions / updates: PIVX and Decred Testnets, BitcoinGreen (cunhasb)
  Monacoin (wakayamap)
* proper generation input handling for various altcoins (erasmospunk) fixing
  `#570`_

Version 1.8.4 (14 Aug 2018)
===========================

* improved notification handling and efficiency
* improved daemon handling with minor fixes; full tests for Daemon class
* remove chain_state class
* various internal cleanups and improvements (erasmospunk)
* add PIVX support (erasmospunk) - mempool handling WIP
* fix protocol 1.3 handling of blockchain.block.header RPC (ghost43)

Version 1.8.3 (11 Aug 2018)
===========================

* separate the DB and the BlockProcessor objects
* comprehensive mempool tests
* fix `#521`_, `#565`_, `#567`_


electrumx/server/env.py
# Copyright (c) 2016, Neil Booth
#
# All rights reserved.
#
# See the file "LICENCE" for information about the copyright
# and warranty status of this software.

'''Class for handling environment configuration and defaults.'''


import re
import resource
from collections import namedtuple
from ipaddress import ip_address

from electrumx.lib.coins import Coin
from electrumx.lib.env_base import EnvBase
import electrumx.lib.util as lib_util


NetIdentity = namedtuple('NetIdentity', 'host tcp_port ssl_port nick_suffix')


class Env(EnvBase):
    '''Wraps environment configuration. Optionally, accepts a Coin class
       as first argument to have ElectrumX serve custom coins not part of
       the standard distribution.
    '''

    # Peer discovery
    PD_OFF, PD_SELF, PD_ON = range(3)

    def __init__(self, coin=None):
        super().__init__()
        self.obsolete(['UTXO_MB', 'HIST_MB', 'NETWORK'])
        self.db_dir = self.required('DB_DIRECTORY')
        self.db_engine = self.default('DB_ENGINE', 'leveldb')
        self.daemon_url = self.required('DAEMON_URL')
        if coin is not None:
            assert issubclass(coin, Coin)
            self.coin = coin
        else:
            coin_name = self.required('COIN').strip()
            network = self.default('NET', 'mainnet').strip()
            self.coin = Coin.lookup_coin_class(coin_name, network)
        self.cache_MB = self.integer('CACHE_MB', 1200)
        self.host = self.default('HOST', 'localhost')
        self.reorg_limit = self.integer('REORG_LIMIT', self.coin.REORG_LIMIT)
        # Server stuff
        self.tcp_port = self.integer('TCP_PORT', None)
        self.ssl_port = self.integer('SSL_PORT', None)
        if self.ssl_port:
            self.ssl_certfile = self.required('SSL_CERTFILE')
            self.ssl_keyfile = self.required('SSL_KEYFILE')
        self.rpc_port = self.integer('RPC_PORT', 8000)
        self.banner_file = self.default('BANNER_FILE', None)
        self.tor_banner_file = self.default('TOR_BANNER_FILE',
                                            self.banner_file)
        self.anon_logs = self.boolean('ANON_LOGS', False)
        self.log_sessions = self.integer('LOG_SESSIONS', 3600)
        # Peer discovery
        self.peer_discovery = self.peer_discovery_enum()
        self.peer_announce = self.boolean('PEER_ANNOUNCE', True)
        self.force_proxy = self.boolean('FORCE_PROXY', False)
        self.tor_proxy_host = self.default('TOR_PROXY_HOST', 'localhost')
        self.tor_proxy_port = self.integer('TOR_PROXY_PORT', None)
        self.peer_discovery_tor = self.peer_discovery_tor_parse()
        # The electrum client takes the empty string as unspecified
        self.donation_address = self.default('DONATION_ADDRESS', '')
        # Server limits to help prevent DoS
        self.max_send = self.integer('MAX_SEND', self.coin.DEFAULT_MAX_SEND)
        self.max_subs = self.integer('MAX_SUBS', 1000000)
        self.max_sessions = self.sane_max_sessions()
        self.max_sessions_per_ip = self.integer('MAX_SESSIONS_PER_IP', 50)
        self.max_sessions_tor = max(self.integer('MAX_SESSIONS_TOR', 1000), 10)  # Require at least 10 for MAX_SESSIONS_TOR
        self.max_session_subs = self.integer('MAX_SESSION_SUBS', 50000)
        self.bandwidth_limit = self.integer('BANDWIDTH_LIMIT', 8000000)
        self.session_timeout = self.integer('SESSION_TIMEOUT', 600)
        self.drop_client = self.custom("DROP_CLIENT", None, re.compile)
        # Blacklist URL. Set this to the empty string in the environment if you want to disable this facility
        self.blacklist_url = self.default('BLACKLIST_URL',
                                          'https://raw.githubusercontent.com/Electron-Cash/electronx-blacklist/master/blacklist.json'
                                          )
        self.blacklist_poll_interval = self.custom('BLACKLIST_POLL_INTERVAL', 300,
                                                   # parse as integer and limit it to >= 30 secs and <= 86400 (1 day)
                                                   lambda x: (int(x) >= 30 and int(x) <= 86400 and int(x)) or 300)
        # If this is set, then we ban whenever an IP address has more than self.max_sessions_per_ip connections
        self.ban_excessive_connections = self.boolean('BAN_EXCESSIVE_CONNECTIONS', True)

        # Identities
        clearnet_identity = self.clearnet_identity()
        tor_identity = self.tor_identity(clearnet_identity)
        self.identities = [identity
                           for identity in (clearnet_identity, tor_identity)
                           if identity is not None]

    def sane_max_sessions(self):
        '''Return the maximum number of sessions to permit.  Normally this
        is MAX_SESSIONS.  However, to prevent open file exhaustion, ajdust
        downwards if running with a small open file rlimit.'''
        env_value = self.integer('MAX_SESSIONS', 1000)
        nofile_limit = resource.getrlimit(resource.RLIMIT_NOFILE)[0]
        # We give the DB 250 files; allow ElectrumX 100 for itself
        value = max(0, min(env_value, nofile_limit - 350))
        if value < env_value:
            self.logger.warning('lowered maximum sessions from {:,d} to {:,d} '
                                'because your open file limit is {:,d}'
                                .format(env_value, value, nofile_limit))
        return value

    def clearnet_identity(self):
        host = self.default('REPORT_HOST', None)
        if host is None:
            return None
        try:
            ip = ip_address(host)
        except ValueError:
            bad = (not lib_util.is_valid_hostname(host)
                   or host.lower() == 'localhost')
        else:
            bad = (ip.is_multicast or ip.is_unspecified
                   or (ip.is_private and self.peer_announce))
        if bad:
            raise self.Error('"{}" is not a valid REPORT_HOST'.format(host))
        tcp_port = self.integer('REPORT_TCP_PORT', self.tcp_port) or None
        ssl_port = self.integer('REPORT_SSL_PORT', self.ssl_port) or None
        if tcp_port == ssl_port:
            raise self.Error('REPORT_TCP_PORT and REPORT_SSL_PORT '
                             'both resolve to {}'.format(tcp_port))
        return NetIdentity(
            host,
            tcp_port,
            ssl_port,
            ''
        )

    def tor_identity(self, clearnet):
        host = self.default('REPORT_HOST_TOR', None)
        if host is None:
            return None
        if not host.endswith('.onion'):
            raise self.Error('tor host "{}" must end with ".onion"'
                             .format(host))

        def port(port_kind):
            '''Returns the clearnet identity port, if any and not zero,
            otherwise the listening port.'''
            result = 0
            if clearnet:
                result = getattr(clearnet, port_kind)
            return result or getattr(self, port_kind)

        tcp_port = self.integer('REPORT_TCP_PORT_TOR',
                                port('tcp_port')) or None
        ssl_port = self.integer('REPORT_SSL_PORT_TOR',
                                port('ssl_port')) or None
        if tcp_port == ssl_port:
            raise self.Error('REPORT_TCP_PORT_TOR and REPORT_SSL_PORT_TOR '
                             'both resolve to {}'.format(tcp_port))

        return NetIdentity(
            host,
            tcp_port,
            ssl_port,
            '_tor',
        )

    def hosts_dict(self):
        return {identity.host: {'tcp_port': identity.tcp_port,
                                'ssl_port': identity.ssl_port}
                for identity in self.identities}

    def peer_discovery_enum(self):
        pd = self.default('PEER_DISCOVERY', 'on').strip().lower()
        if pd in ('off', ''):
            return self.PD_OFF
        elif pd == 'self':
            return self.PD_SELF
        else:
            return self.PD_ON

    def peer_discovery_tor_parse(self):
        tpd = self.default('PEER_DISCOVERY_TOR', 'off').strip().lower()  # if 'on', we forward .onion peers we hear about. otherwise off (due to sybil attack issues)
        if tpd in ('on', 'yes', 'true', '1', 'enabled', 'enable'):
            return True
        return False


    def get_info(self):
        ''' Used by rpc_getenv to get all the env vars we are using. '''
        env = {}
        for name in dir(self).copy():
            if name.startswith('_'):
                continue
            value = getattr(self, name, None)
            if not isinstance(value, (str, int, float, bool)):
                continue
            if isinstance(value, bool):
                value = "1" if value else ""  # Map bools to non-empty/empty strings to match how they are parsed
            env[name.upper()] = value
        # now mogrify some values
        env['PEER_DISCOVERY'] = {
            self.PD_ON : "on", self.PD_OFF : "off", self.PD_SELF : "self"
        }.get(self.peer_discovery, 'on')
        env['PEER_DISCOVERY_TOR'] = "on" if self.peer_discovery_tor else "off"
        clear_id = self.clearnet_identity()
        tor_id = self.tor_identity(clear_id)
        if tor_id is not None:
            env['REPORT_HOST_TOR'] = str(tor_id.host)
            env['REPORT_TCP_PORT_TOR'] = str(tor_id.tcp_port)
            env['REPORT_SSL_PORT_TOR'] = str(tor_id.ssl_port)
        if clear_id is not None:
            env['REPORT_HOST'] = str(clear_id.host)
            env['REPORT_TCP_PORT'] = str(clear_id.tcp_port)
            env['REPORT_SSL_PORT'] = str(clear_id.ssl_port)
        return env
